### PR TITLE
Allow edits to popup container class while the popup is not on the map

### PR DESCRIPTION
**Pull request by SnailBones**
Currently, `addClassName`, `removeClassName` and `toggleClassName` all have no effect when called before the popup is added to the map, and all have an incorrect example.

In this P.R. I add a `Popup._classList` Set that keeps track of the added classes, pushing them to the DOM whenever the popup is added to the map.

At @arindam1993 and @rreusser's [suggestion](https://github.com/mapbox/mapbox-gl-js/pull/10889#issuecomment-884502015), the popup now uses a Set to keep track of its classes, rewriting the element classList on every update to the popup. This creates a single source of truth and prevents strange behavior when toggling a default class. However, for users who are bypassing the methods to directly manipulate the element classList, this is a breaking change.

Other changes:

- Returns for `addClassName` and `removeClassName` to allow for method chaining.
- Since refactoring, popups no longer throw errors for addinglremoving/toggling invalid strings. (Classes with whitespace instead add multiple classes to the DOM). Lines checking for errors in tests have been removed.
- Moved class maniuplation function from anchor.js into marker.js since it is no longer shared by popup.js.

## Launch Checklist

 - [x] Allow `addClassName` to be called when popup is not on map.
 - ~~Add warnings for incorrect user actions (calling `removeClassName` and `toggleClassName` before popup is on map).~~
 -  ~~Fix examples for `removeClassName` and `toggleClassName`~~ 
 - [x] Allow `removeClassName` and `toggleClassName` to be called when popup is not on map.
 - [x] Tests for `addClassName`.
 - [x]  Tests for `removeClassName` and `toggleClassName`.
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Popup.addClassName, removeClassName and toggleClassName work while the popup is not on the map, and no longer throw DOM errors for invalid class names. Directly modifying the popup container CSS class no longer works, all edits should use these functions. </changelog>`
 - [x] Implement Set with one-way update.

### CODE REVIEW

File: src/ui/popup.js [hunk 1]
```diff
@@ -443,7 +443,9 @@ export default class Popup extends Evented {
     addClassName(className: string) {
         if (this._container) {
             this._container.classList.add(className);
-        }
+        } else if (this.options.className && this.options.className !== ``) {
```
Comment (rreusser) [inline src/ui/popup.js:446]: Nit: I believe `''` is falsey so that it's impossible to pass the first condition but not the second.

```suggestion
        } else if (this.options.className) {
```

File: src/ui/popup.js [hunk 2]
```diff
@@ -452,12 +454,17 @@ export default class Popup extends Evented {
      * @param {string} className Non-empty string with CSS class name to remove from popup container
      *
      * @example
-     * let popup = new mapboxgl.Popup()
-     * popup.removeClassName('some-class')
+     * const popup = new mapboxgl.Popup({className: 'some classes'})
+        .setText('Popup')
```
Comment (rreusser) [inline src/ui/popup.js:458]: Nit: I think the asterisks are necessary to continue a jsdoc comment. Same in block below

```suggestion
    *  .setText('Popup')
```

File: src/ui/popup.js [hunk 3]
```diff
@@ -482,11 +489,18 @@ export default class Popup extends Evented {
      *
      * @example
      * let popup = new mapboxgl.Popup()
-     * popup.toggleClassName('toggleClass')
+     * const popup = new mapboxgl.Popup({className: 'highlighted'})
+        .setText('Popup')
```
Comment (rreusser) [inline src/ui/popup.js:493]: Nit: ditto on asterisks.

File: src/ui/popup.js [hunk 4]
```diff
@@ -482,11 +489,18 @@ export default class Popup extends Evented {
...
+        .setLngLat([0, 0])
+        .addTo(map);
+     * popup.toggleClassName('highlighted')
+     * popup.toggleClassName('highlighted')
      */
     toggleClassName(className: string) {
         if (this._container) {
             return this._container.classList.toggle(className);
+        } else {
```
Comment (rreusser) [inline src/ui/popup.js:502]: Super nitpicky 🙈 : can omit `} else {` with early return.

```suggestion
        }
```

**Comment by arindam1993** [comment]
I'd be a fan of moving the `classList` out from the container and onto an array(or `Set` maybe? 🤔 ) on the Popup instance itself.
That way we could lift the restriction on remove and toggle as well, and implement a one-way update form `this.classList -> this._container.classList` whenever changes are triggered.

**Comment by karimnaaji** [comment]
@SnailBones I believe the failed CI tests are unrelated to your PRs. I would pull in @arindam1993 's change from yesterday either rebasing or merging from latest main, as you prefer.

**Comment by SnailBones** [comment]
> I'd be a fan of moving the `classList` out from the container and onto an array(or `Set` maybe? 🤔 ) on the `Popup` instance itself.
> That way we could lift the restriction on remove and toggle as well, and implement a one-way update form `this.classList -> this._container.classList` whenever changes are triggered.

@arindam1993 What do you think of the implementation with another DOM element [here](https://github.com/mapbox/mapbox-gl-js/pull/10143)?

**Review by rreusser — CHANGES_REQUESTED** [review]
Leaving some super nitpicky details as a review, but I rather like @arindam1993's suggestion of using a `Set` here. It's not without precedence in GL JS, though I'm not 100.0% sure this usage was definitely intentional and didn't just happen to sneak through:

https://github.com/mapbox/mapbox-gl-js/blob/780a0bb1ded8f80024f72ecacbe29abf75e3dc10/src/util/mapbox.js#L573

I guess spread with `...` works as well for concatenating…

```js
classList = new Set()
// Set(0) {}
classList.add('foo')
// Set(1) {"foo"}
classList.add('bar')
// Set(2) {"foo", "bar"}
[...classList].join(' ')
// "foo bar"
```

File: src/ui/popup.js (hunk 2 above)
Comment (SnailBones) [reply src/ui/popup.js:458]: Good catch!

**Comment by SnailBones** [comment]
> ```js
> classList = new Set()
> // Set(0) {}
> classList.add('foo')
> // Set(1) {"foo"}
> classList.add('bar')
> // Set(2) {"foo", "bar"}
> [...classList].join(' ')
> // "foo bar"
> ```
@rreusser That's a clever implementation! I'm still struggling to understand what the use case is though. Do you have an example of when it might be necessary to manipulate classes in a popup before it has been added to the map?

**Comment by rreusser** [comment]
Ah, I wasn't 100% certain whether we wanted to preemptively create all of the DOM elements (I imagine they're relatively lightweight, but how many would people create?), so I thought maybe instead of lifting the DOM element's `classList` onto an instance property, we could maybe just track the classList as a set, then concatenate and set it on an inserted DOM element directly whenever it changes.

https://github.com/mapbox/mapbox-gl-js/pull/10143/files#diff-de43205af4d4066c2e22519c9f5fa5e52ff160ed43091938190addbe4b46f4a0R117

**Comment by ryanhamley** [comment]
I can imagine a situation in which popups are set up with a default theme and users can change that theme before they're added to the map, which would make `remove` and `toggle` useful. I agree with the idea of maintaining a class list per popup instance that can be updated without a DOM element and then the DOM element can be updated as is if it exists. IMO, we should avoid creating DOM elements unless absolutely necessary since DOM creation can be time- and memory-consuming.

File: src/ui/popup.js [hunk 5]
```diff
@@ -435,30 +437,36 @@ export default class Popup extends Evented {
      * Adds a CSS class to the popup container element.
      *
      * @param {string} className Non-empty string with CSS class name to add to popup container
+     * @returns {Popup} `this`
```
Comment (SnailBones) [inline src/ui/popup.js:440]: Added a return to allow chaining

File: src/ui/popup.js [hunk 6]
```diff
@@ -107,11 +107,13 @@ export default class Popup extends Evented {
     _lngLat: LngLat;
     _trackPointer: boolean;
     _pos: ?Point;
+    _classList: Set<string>;
 
     constructor(options: PopupOptions) {
         super();
         this.options = extend(Object.create(defaultOptions), options);
         bindAll(['_update', '_onClose', 'remove', '_onMouseMove', '_onMouseUp', '_onDrag'], this);
+        this._classList = new Set(options && options.className ? options.className.split(` `) : []);
```
Comment (rreusser) [inline src/ui/popup.js:116]: Maybe trim and split by one _or more_ whitespace to increase robustness (e.g. silly input like `"  foo   bar "`).

```suggestion
        this._classList = new Set(options && options.className ? options.className.trim().split(/\s+/) : []);
```

File: src/ui/popup.js [hunk 7]
```diff
@@ -435,30 +437,36 @@ export default class Popup extends Evented {
...
+     * @returns {Popup} `this`
      *
      * @example
      * let popup = new mapboxgl.Popup()
      * popup.addClassName('some-class')
      */
     addClassName(className: string) {
+        this._classList.add(className);
         if (this._container) {
             this._container.classList.add(className);
```
Comment (rreusser) [inline src/ui/popup.js:449]: What do you think about overwriting classes with something like `this._applyClassList()` which would look like:

```js
_applyClassList () {
  if (!this._container) return;
  this._container.className = [...this._classList.values()].join(' ');
}
```

My main concern is that we have two independent sources of truth which we _hope_ remain synchronized. But since we don't allow a user-supplied container anyway, so I don't _think_ there should be a case where this would prevent you from accomplishing some behavior you couldn't simply achieve by using the supplied API.

However, this could be a breaking change if someone was previously manipulating the container class directly.

File: src/ui/popup.js [hunk 8]
```diff
@@ -521,8 +535,8 @@ export default class Popup extends Evented {
             this._container = DOM.create('div', 'mapboxgl-popup', this._map.getContainer());
             this._tip       = DOM.create('div', 'mapboxgl-popup-tip', this._container);
             this._container.appendChild(this._content);
-            if (this.options.className) {
-                this.options.className.split(' ').forEach(name =>
+            if (this._classList) {
+                this._classList.forEach(name =>
```
Comment (rreusser) [inline src/ui/popup.js:539]: If we overwrote classes, then I think `this._container.className = [...this._classList.values()].join(' ')` would be the simplest way to accomplish this.

**Review by rreusser — COMMENTED** [review]
Looks good! I'm somewhere between "request changes" and "approve" since the PR as-is looks good to me, but if it's reasonable to remove a source of truth and simply overwrite the container classes whenever they change, that would be my top preference, if only to avoid a potential corner case in which they get out of sync.

File: src/ui/popup.js (hunk 6 above)
Comment (SnailBones) [reply src/ui/popup.js:116]: Good call!

File: src/ui/popup.js (hunk 5 above)
Comment (ryanhamley) [reply src/ui/popup.js:440]: Nit: During the doc's sprint, we updated all returns of `this` to be:

`@returns {...} Returns itself to allow for method chaining.`

File: src/ui/popup.js (hunk 7 above)
Comment (SnailBones) [reply src/ui/popup.js:449]: I don't think there's a _valid_ reason for anyone to manipulate the container class directly right now. (Still doesn't mean that no one's doing it.)

It wouldn't be quite as simple as the code you propose: the popup DOM element has additional default classes. So this:
```js
 const popup = new Popup({className: 'some classes'});
 ```
creates this element:
```html
 <div class="mapboxgl-popup mapboxgl-popup-anchor-bottom yellow"></div>
```

**Review by rreusser — APPROVED** [review]
Nice work on this, @SnailBones! Switching my review to approve. 👍 👏

File: src/ui/popup.js (hunk 7 above)
Comment (rreusser) [reply src/ui/popup.js:449]: Good point! Just wanted to be sure.

File: src/ui/popup.js [hunk 9]
```diff
@@ -482,12 +490,18 @@ export default class Popup extends Evented {
...
      * let popup = new mapboxgl.Popup()
-     * popup.toggleClassName('toggleClass')
+     * popup.toggleClassName('highlighted')
      */
     toggleClassName(className: string) {
         if (this._container) {
-            return this._container.classList.toggle(className);
+            this._container.classList.toggle(className);
+        }
+        if (this._classList.has(className)) {
```
Comment (arindam1993) [inline src/ui/popup.js:499]: I think you can directly use `delete` here because it returns `true` if the entry existed in the set and was successfully removed.

File: src/ui/popup.js (hunk 7 above)
Comment (SnailBones) [reply src/ui/popup.js:450]: I did find one weird edge case: if the user toggles one of the default classes:

`popup.toggleClassName('mapboxgl-popup');`

This removes the item from the container while adding it to the list.

I'm sure we could fix this, but I'm not sure what the correct behavior would be or if it's worth the lift and added complexity just to solve a very specific edge case.

File: src/ui/popup.js (hunk 9 above)
Comment (SnailBones) [reply src/ui/popup.js:500]: Can you elaborate on what you mean by that @arindam1993?
Are you suggesting changing the following line to:
```js
return this.classList.delete(className);
```

File: test/unit/ui/popup.test.js [hunk 10]
```diff
@@ -599,14 +599,74 @@ test('Popup adds classes from className option, methods for class manipulations
...
+    popup.removeClassName('some');
+    popupContainer = popup.addTo(map).getElement();
+
+    t.ok(!popupContainer.classList.contains('some'));
+
+    popup.remove();
+    popup.toggleClassName('toggle');
+    popupContainer = popup.addTo(map).getElement();
 
-    t.throws(() => popup.addClassName(''), window.DOMException);
```
Comment (SnailBones) [inline test/unit/ui/popup.test.js:606]: Since DOM is no longer directly updated, these functions no longer throw an error.

File: src/ui/popup.js [hunk 11]
```diff
@@ -600,10 +619,9 @@ export default class Popup extends Evented {
 }
 
 function normalizeOffset(offset: ?Offset) {
-    if (!offset) {
-        return normalizeOffset(new Point(0, 0));
```
Comment (SnailBones) [inline src/ui/popup.js:604]: Seemed like a strange case for recursion here

**Comment by karimnaaji** [comment]
@SnailBones You added a 'breaking change' label to this PR, is this breaking only in the way that popups were throwing errors on adding/removing/toggling and not anymore or is it something else?

If it's breaking and requires user code changes, is there any transition that our users should follow once we release concerning this behavior? I'm interested to know whether this warrants an extra note on the CHANGELOG.

**Comment by SnailBones** [comment]
Two breaking changes.

First as you mention, the lack of throwing errors on invalid popup class names.

Also, if a user has directly modified classList/className instead of using the popup functions, this change means that updates will reset classes added or removed in this way. The fix here would be ensuring that all modifications to the CSS class are performed with the three functions.

I think that both of these breaking changes should only affect very rare edge cases. The former could be an issue if an app allowed for custom popup CSS classes (I can't think of a good reason for that, but maybe there's a reason it was in the unit tests.) The latter is something that users shouldn't be doing, but I don't think we can say with certainty that they're not.

**Comment by SnailBones** [comment]
Just updated the changelog to:
<changelog>Popup.addClassName, removeClassName and toggleClassName work while the popup is not on the map, and no longer throw DOM errors for invalid class names. Directly modifying the popup container CSS class no longer works, all edits should use these functions. </changelog>

File: src/ui/popup.js (hunk 9 above)
Comment (arindam1993) [reply src/ui/popup.js:490]: No, i meant 
```
if(this.classList.delete(className)){
   finalState = false; 
} else {
   finalState = true;
   this._classList.add(className);
}
```

I think `Set#delete` already internally does a `Set#has` check, so I was suggesting leveraging  that

**Review by arindam1993 — APPROVED** [review]
Nice cleanup @SnailBones. 
Just one more suggestion, otherwise its a 🟢  from me!
I don't think this counts as a breaking change since it would be a bit weird if someone was actually relying on classnames getting lost.

File: src/ui/popup.js [hunk 12]
```diff
@@ -512,6 +522,47 @@ export default class Popup extends Evented {
...
+        if (this.options.anchor) { return this.options.anchor; }
+
+        const pos: any = this._pos;
+        const width = this._container.offsetWidth;
+        const height = this._container.offsetHeight;
+
+        let anchorComponents;
+        if (pos.y + offset.bottom.y < height) {
+            anchorComponents = ['top'];
+        } else if (pos.y > this._map.transform.height - height) {
```
Comment (rreusser) [inline src/ui/popup.js:535]: What does it mean when `this._map.transform.height` does not equal `_container.offsetHeight`?

File: src/ui/popup.js [hunk 13]
```diff
@@ -512,6 +522,47 @@ export default class Popup extends Evented {
...
+    _updateClassList() {
+        const classes = new Set(this._classList);
+        classes.add('mapboxgl-popup');
+        if (this._anchor) {
+            classes.add(`mapboxgl-popup-anchor-${this._anchor}`);
+        }
+        if (this._trackPointer) {
+            classes.add('mapboxgl-popup-track-pointer');
+        }
+        this._container.className = [...classes].join(' ');
```
Comment (rreusser) [inline src/ui/popup.js:563]: I think this could be rearranged to reduce a set allocation (spread to arrray, then `.push` optional classes, then join), but it's so far down in the list of memory usage, that I fully support the above 👍

File: src/ui/popup.js [hunk 14]
```diff
@@ -539,46 +582,22 @@ export default class Popup extends Evented {
             this._lngLat = smartWrap(this._lngLat, this._pos, this._map.transform);
         }
 
-        if (this._trackPointer && !cursor) return;
+        if (!this._trackPointer || cursor) {
```
Comment (rreusser) [inline src/ui/popup.js:585]: Nit: Is there a reason for inverting the conditional? Early return can be nice to de-nest code when otherwise equivalent. Nothing blocking here for me though.

File: src/ui/popup.js (hunk 11 above)
Comment (rreusser) [reply src/ui/popup.js:604]: I believe I've seen this style, especially in the symbol code, where it runs modified values through a different branch of the same function, but TBH I find it slightly hard to follow, so that if otherwise equivalent, I prefer your change 👍

**Review by rreusser — APPROVED** [review]
Looks great! Despite having pointed out a couple nits, this looks great, and I'd opt to merge as-is.

I think @arindam1993 is probably right that we're fine not considering this a breaking change. In some sense, [literally any fix is breaking if people relied upon broken behavior](https://m.xkcd.com/1172/), so it's probably safe to explain it clearly in the changelog but consider it really just a fix for something people may have been doing which would have had undefined results.

File: src/ui/popup.js (hunk 9 above)
Comment (SnailBones) [reply src/ui/popup.js:490]: Good idea!

File: src/ui/popup.js (hunk 12 above)
Comment (SnailBones) [reply src/ui/popup.js:535]: I'm not sure what you're getting at here. `this._map.transform.height` is the height of the map, while `_container.offsetHeight` is the height of the popup.

File: src/ui/popup.js (hunk 13 above)
Comment (SnailBones) [reply src/ui/popup.js:563]: Good idea!

File: src/ui/popup.js (hunk 14 above)
Comment (SnailBones) [reply src/ui/popup.js:585]: My thinking was to remove the early return allowing `this. _updateClassList()` to be called in any condition.

File: src/ui/popup.js (hunk 12 above)
Comment (rreusser) [reply src/ui/popup.js:535]: Ohhhh sorry. I was thinking map container element. 🤦

File: src/ui/popup.js (hunk 14 above)
Comment (SnailBones) [reply src/ui/popup.js:584]: Yup it was redundant! (Oops) Fixed in latest commit.
